### ISSUE TITLE
aead encryption/decryption turns nul 0's into ASCII 0 chars

### ISSUE THREAD

**Issue by BlackHoleFox** [briansmith/ring#753]
I noticed that when using the `aead::seal_in_place` and `aead::open_in_place` functions, it turns any nul 0 bytes into ASCII 0 characters with a hex value of `30` instead of the original `00` they were in the vector before sealing it.

Code for encryption (blobdefault is a string and filldata is a Vec\<u8>): 
```rust
let mut filldata = blobdefault.into_bytes();

filldata.push(0);
filldata.push(0);
aead::seal_in_place(&cipherkey, &nonce, &[0; 0], &mut filldata, aead_algo_tag_len).unwrap();
```
A data dump of `filldata` in a hex editor reveals it as expected, with the proper `00` values:

![image](https://user-images.githubusercontent.com/20936452/50677002-52aa3e80-0fef-11e9-81ce-d745bb04b892.png)

After decrypting the data using the `aead:open_in_place()` and dumping the decrypted data into a file the hex editor shows the 0s as hex value `30`s now:

![image](https://user-images.githubusercontent.com/20936452/50677025-78374800-0fef-11e9-97b2-f01be64b3cac.png)

Is this a intentional thing the encryption or decryption function does or an unintended bug? Thanks!

Ring version in `Cargo.toml`: `ring = "0.13.5"`
Rust toolchain: `stable-x86_64-pc-windows-msvc | rustc 1.31.1`

**Comment by BlackHoleFox** [comment]
It sounds like I may of worded that in a confusing way. Its not the `\r\n` that is the issue presently, its the highlighted 0's that I pushed to the vector that are separate from the newline, even though they both show up as `..` in the hex editor. I can get a code sample if needed though if wrong about the miscommunication.

**Comment by briansmith** [comment]
Try `aead::seal_in_place(&cipherkey, &nonce, &[0; 0], &mut filldata, aead_algo_tag_len + 2).unwrap();`

This indicates that you want the tag to be written 2 bytes from the end `filldata`. This will leave the last two zero bytes unchanged, and the plaintext from `open_in_place` will not have the spaces at  the end.

To understand better what is happening, instead of having "foxing\r\n" followed by 16 spaces followed by two zero bytes, try an input like "foxing\r\nabcdefghijklmnopqr". If you change the input string like that, then the output of `open_in_place()` in your original program will be "foxing\r\nab". If you change your original program to use `aead::seal_in_place(&cipherkey, &nonce, &[0; 0], &mut filldata, aead_algo_tag_len + 2).unwrap();` then you will get "foxing\r\n".

**Comment by BlackHoleFox** [comment]
Im intending for the 2 `0`'s to be part of the ciphertext though, not to remain unchanged. I'm adding them to filldata as extra bytes right after the actual input data, so the ten encrypted bytes is correct, I think. Here's a gist if it helps any. Code snippets should help more then words here: https://gist.github.com/BlackHoleFox/8e2b3e9c91329a8b6d474798ace9314b

And the dumps that I get running my code myself: [Data_Dumps.zip](https://github.com/briansmith/ring/files/2726366/Data_Dumps.zip)

**Comment by briansmith** [comment]
> Im intending for the 2 `0`'s to be part of the ciphertext though, not to remain unchanged. I'm adding them to filldata as extra bytes right after the actual input data, so the ten encrypted bytes is correct, I think.

No, you're adding them after the 16 bytes you reserved for the tag. `seal_in_place` will always put the 16-byte tag at the very end of the buffer you gave it, overwriting 14 0x30 bytes plus your two 0x00 bytes with the tag, leaving two 0x30 bytes before the tag.

If you want them to be right after the `\r\n` then you need to push them *before*, not *after* you push the 16 bytes of '0'.

**Comment by BlackHoleFox** [comment]
Ohhh. Makes sense now why it wasn't working. Sorry for all the trouble. Thank you for the explanation into why it wasn't working though.
